automata/fa/nfa.py

```diff
@@ -1009,3 +970,89 @@ def add_any_transition(start_state_dict, end_state):
             initial_state=(0, 0),
             final_states=final_states,
         )
+
+    def iter_transitions(
+        self,
+    ) -> Generator[Tuple[NFAStateT, NFAStateT, str], None, None]:
+        return (
+            (from_, to_, symbol)
+            for from_, lookup in self.transitions.items()
+            for symbol, to_lookup in lookup.items()
+            for to_ in to_lookup
+        )
+
+    def _get_input_path(
```

Comment:
@khoda81 it looks like this function fails on some inputs (see the `test_nfa.py` file and the changes in the latest commit for details https://github.com/caleb531/automata/pull/129/commits/6f1ec9fd7086e9ba7b5d65eb459ae627ae9cfcea). It also looks like we can't use `functools.cache` in Python 3.8, so it would be great to modify this function to remove that usage and fix whatever bugs we're running into.

Reply:
@eliotwrobson @khoda81 On the note about the caching, `functools.lru_cache(maxsize=None)` seems to be the way to go, since it was added back in Python 3.2, and the docs clearly state it is the same as `functools.cache`:

> Returns the same as lru_cache(maxsize=None)
Source: https://docs.python.org/3/library/functools.html#functools.cache

I will go ahead and make this change, since it should only take 2 minutes.

Reply:
@eliotwrobson @khoda81 Okay, change is made—please remember to pull.

Also, I accidentally resolved the conversation, but I unresolved it now since there is still the matter of behavioral correctness to discuss/resolve.

Reply:
Rewrote this algorithm to fix the correctness issue! https://github.com/caleb531/automata/pull/129/commits/10706a6406d9057a727e2ab49371a3f0b354c009